### ISSUE TITLE
CSP too strict for certain custom logout URIs

### ISSUE THREAD

**Issue by tjmv** [firefly-iii/firefly-iii#5493]
### Support guidelines

- [X] I've read the [support guidelines](https://github.com/firefly-iii/firefly-iii/blob/main/.github/support.md)

### I've found a bug and checked that ...

- [X] ... [the documentation](https://docs.firefly-iii.org/) does not mention anything about my problem
- [X] ... there are no open or closed issues that are related to my problem
- [X] ... it's [definitely me, not you](https://github.com/firefly-iii/firefly-iii/blob/main/.github/its_you_not_me.md)

### Description

I have found myself unable to logout of Firefly III when using a remote authentication guard configured on a different subdomain (using domain level mode authentication). The browser's console returns the following error:
`Refused to send form data to 'https://<ff subdomain>/logout' because it violates the following Content Security Policy directive: "form-action 'self'".`

Visually it's as if nothing happens when clicking the log out action, but it loads a redirect that's being ignored.

### Expected behaviour

The browser sends me to the configured logout page.

### Debug information

Debug information generated at 2022-01-05 22:43:01 Europe/Amsterdam for Firefly III version **5.6.9**.

| Scope | Version |
| --- | --- |
| Firefly III | 5.6.9 |
| Firefly III API | 1.5.5 |
| Build | 66, 02-01-2022 08:12:35 CET |
| PHP | 8.0.13 |
| Host | Linux |

| System info | Value |
| --- | --- |
| Installation ID | 779b8124-8924-48e2-9720-b40593ad71d2 |
| Using docker? | true |
| Layout | v1 |
| System TZ | Europe/Amsterdam |
| Browser TZ | Europe/Amsterdam |
| App environment | local |
| App debug mode | true |
| App cache driver | redis |
| App logging | debug, stack |
| Display errors | Off |
| Error reporting | ALL errors |
| Interface | apache2handler |
| Default language | en_US |
| Default locale | equal |
| BCscale | 24 |
| DB drivers | mysql, pgsql, sqlite |
| Current driver | pgsql |
| DB version | 18 (exp. 18) |
| Login provider | eloquent  |
| Trusted proxies (.env) | ** |

| User info | Value |
| --- | --- |
| Session start | 2022-01-01 00:00:00 |
| Session end | 2022-01-31 23:59:59 |
| Session first | 2022-01-01 00:00:00 |
| User ID | 1 |
| User language | nl_NL |
| User locale | nl_NL |
| Attempt at "nl_NL.utf8" | 'nl_NL.utf8' |
| Attempt at "nl_NL.UTF-8" | 'nl_NL.UTF-8' |
| User agent | Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/96.0.4664.110 Safari/537.36 Edg/96.0.1054.62 |

### Steps to reproduce

1. Configure a remote authentication guard
2. Set-up a custom logout URI on a different (sub-)domain
3. Login to Firefly III using the remote authentication guard
4. Logout using the interface.

### Additional info

A quick test to include the custom logout URI in the form-action CSP fixed the issue:
```diff
--- app/Http/Middleware/SecureHeaders.old.php	Wed Jan  5 22:46:47 2022
+++ app/Http/Middleware/SecureHeaders.new.php	Wed Jan  5 22:46:47 2022
@@ -64,7 +64,12 @@
 
         $route = $request->route();
         if (null !== $route && 'oauth/authorize' !== $route->uri) {
-            $csp[] = "form-action 'self'";
+            $authGuard = config('firefly.authentication_guard');
+            $logoutUri = '';
+            if ('remote_user_guard' === $authGuard) {
+                $logoutUri = config('firefly.custom_logout_url');
+            }
+            $csp[] = sprintf("form-action 'self' %s", $logoutUri);
         }
 
         $featurePolicies = [
```

**Comment by JC5** [comment]
It should redirect, not POST that URL. I'll pick it up!

**Comment by tjmv** [comment]
Unfortunately, I don't think you can easily change the HTTP request method to GET. If the browser receives a redirect, it'll use the same method. Moreover, in my opinion a POST request makes more sense for these actions as well, but I cannot find any specifications on how to deal with logging out when dealing with these remote authentication methods. If an IdP or whatever the endpoint is does not accept a POST request, then -- in my opinion -- that's not a problem FF has to deal with. The bigger OAuth2 providers do use an explicit session id / token in their requests (both accepting GET and POST) with an optional return URI, but the popular forward auth implementations I've looked at simply set a cookie and don't include the relevant info in the headers. So I do think there should be an option to specify more logout parameters, but that's another issue.

Making the logout send POST request in seems to be done by the client in Firefly III, so changing it to GET in only certain situations would involve changing all relevant templates and adding some logic there or have a dirty html/javascript redirect on a separate logout page.

As for my included CSP change, I simply looked at the [MDN docs on the topic](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/form-action#sources) and it suggests a full URI is valid, though maybe it should be encoded first to avoid spaces. This is by far the easiest solution I could think of and it works fine.

**Comment by JC5** [comment]
Thanks for the expansive reply. The old behavior until recently was a simple redirect header towards the logout URL. I can restore that with no problems.

> that's not a problem FF has to deal with. 

So far, the endpoints that required POST were the problems I did not have to deal with 😉

**Comment by JC5** [comment]
I don't get it. I've been testing this under the assumption I programmed it to POST you to `https://logouturl/`. But that's not how it works.

It will always POST to `your-firefly/logout` and THEN redirect you. And it works, even with entirely different URL's.

So not sure what the issue is exactly.

**Comment by tjmv** [comment]
I checked it in Firefox and there it's exactly as you describe. However, all the Chromium based browsers refuse to load it with the current content policy.

**Comment by tjmv** [comment]
While I mentioned POST headers before, apparently I got some of it wrong. I retested it again and the process is as follows:

1. Open any page on firefly-iii so that it loads the csp with `form-action 'self'` and has a visible logout button.
2. Press the logout button to make a POST request to /logout
3. The page is processed and returns a response with HTTP status 302, location and the regular headers. The form-action csp header doesn't matter at this point.
4a. In Chromium based browsers the request is blocked now. However, when it's not blocked the process is identical to Firefox's.
4b. In Firefox a GET request is made to the configured URI.

So just one POST request followed by a GET to the external host if all goes well. Must've read the wrong headers or some debug stuff I inserted caused a different result. My apologies for the confusion.
